CloudMap support for spring aws applications

hariohmprasath opened the pull request:
## :loudspeaker: Type of change
<!--- Put an `x` in the boxes that apply -->
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring


## :scroll: Description
<!--- Describe your changes in detail -->
This pull request adds support for integrating spring boot application with AWS cloudmap. Here is an example of how an simple yaml file integration looks like for both cloudmap registration and discovery

```yaml
aws:
    cloudmap:
        discovery:
            discoveryList:
            -   healthCheckProtocol: http
                healthCheckResourcePath: /health
                healthCheckThreshold: 5
                service: test-service
                serviceNameSpace: test-namespace
            failFast: true
        enabled: true
        region: us-east-1
        registry:
            description: Namespace for sample cloudmap registry service
            port: 80
            service: service-service
            serviceNameSpace: service-namespace
```

You can also use annotations to register the spring boot applications to cloudmap

```java
@CloudMapRegistry(
		serviceNameSpace = "dark1-namespace",
		service = "dark1-service",
		port = 80
)
public class TestController {}
```
> **Limitation:** This pull requests support cloudmap discovery on both EC2 and Fargate based ECS instances, but for service registry currently it only supported for Fargate instance as it uses the`ECS_CONTAINER_METADATA_URI_V4` URL. 

## :bulb: Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Adds support for cloudmap integration with spring boot applications for both service registry and discovery

<!--- If it fixes an open issue, please link to the issue here. -->
#5 

## :green_heart: How did you test it?
I executed the following tests:
* Ran unit test cases and made sure nothing failed
* Deployed the sample spring boot app checked in part of this pull request in ECS fargate instance. The application could successfully discover and register the application to AWS cloudmap.


## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->
- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] I updated reference documentation to reflect the change
- [x] All tests passing
- [x] No breaking changes


## :crystal_ball: Next steps
* Complete code review
* Add support for cloudmap registration from EC2 instances

eddumelendez reviewed (changes requested):
@hariohmprasath thank you so much for the PR! I am really excited about this one TBH and I don't know why. I have submitted some comments to help you organize the modules much better. Please, don't forget to add yourself as an author for the classes that have been touched.

I'm not an expert in cloudmap and left one question to be clear about it. Also, it would be nice to add some javadocs to know more the class responsibility.

spencergibb commented on spring-cloud-aws-cloudmap-config/src/main/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapPropertySource.java:
```diff
@@ -0,0 +1,86 @@
+
+import org.springframework.core.env.EnumerablePropertySource;
+
+public class AwsCloudMapPropertySource extends EnumerablePropertySource<AWSServiceDiscovery> {
```
I second the question

hariohmprasath replied:
No problem happy to explain. Cloudmap has two parts to it.
* Registration --> Where applications can register itself to CloudMap service
* Discovery --> Where application can discover other registered instances using namespace & service name. We are using property store to allow discovery of these instances.

Here is an example of a property getting injected with HttpInstanceSummary (in string format) from cloudmap service based on CloudMap namespace and service name using the key "namespace/service"

```java
@Value("${test-namespace/test-service}")
private String registryDetails;
```

You can find this example of this inside [SpringCloudAwsCloudMapSample.java](spring-cloud-aws-samples
/spring-cloud-aws-samples/src/main/java/io/awspring/cloud/cloudmap/sample)

Hope this answers your questions, if you have any further questions, let me know. Happy to explain.

eddumelendez replied:
Thanks for the explanation @hariohmprasath. However, I think we should give this more a flavor similar to [spring-cloud-consul-discovery](https://github.com/spring-cloud/spring-cloud-consul/tree/main/spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul) module which brings the service registry and discovery.

hariohmprasath commented on spring-cloud-aws-cloudmap-config/src/main/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapDiscovery.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.util.List;
+
+public class AwsCloudMapDiscovery {
```
Updated. Thanks

hariohmprasath commented on spring-cloud-aws-cloudmap-config/pom.xml:
```diff
@@ -0,0 +1,61 @@
+		<version>2.3.0-RC2</version>
+	</parent>
+
+	<artifactId>spring-cloud-aws-cloudmap-config</artifactId>
```
Updated. Thanks

hariohmprasath commented on spring-cloud-aws-cloudmap-config/pom.xml:
```diff
@@ -0,0 +1,61 @@
+
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-configuration-processor</artifactId>
```
Removed from dependencies

hariohmprasath commented on spring-cloud-aws-cloudmap-config/src/main/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapDiscoveryProperties.java:
```diff
@@ -0,0 +1,65 @@
+
+import java.util.Map;
+
+public class AwsCloudMapDiscoveryProperties {
```
Updated

hariohmprasath commented on spring-cloud-aws-cloudmap-config/src/main/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapDiscoveryProperties.java:
```diff
@@ -0,0 +1,65 @@
+	private Map<String, String> filterAttributes;
+
+	public String getServiceNameSpace() {
+		return serviceNameSpace;
```
Updated.

hariohmprasath commented on spring-cloud-aws-cloudmap-config/src/main/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapDiscoveryProperties.java:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	public String getService() {
+		return service;
```
Updated.

hariohmprasath commented on spring-cloud-aws-cloudmap-config/src/main/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapProperties.java:
```diff
@@ -0,0 +1,86 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@ConfigurationProperties(AwsCloudMapProperties.CONFIG_PREFIX)
+
```
Removed

hariohmprasath commented on spring-cloud-aws-cloudmap-config/src/main/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapProperties.java:
```diff
@@ -0,0 +1,86 @@
+
+@ConfigurationProperties(AwsCloudMapProperties.CONFIG_PREFIX)
+
+public class AwsCloudMapProperties {
```
I followed the same model that we have for parameter store, where we have a module for reading config and a starter project for beans and initiation. Is it ok if I keep it the same way? Or do you want me to move it here? Because this might require a lot more refactoring, so thought of checking with you before that.

hariohmprasath commented on spring-cloud-aws-cloudmap-config/src/main/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapProperties.java:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	public String getRegion() {
+		return region;
```
Updated

hariohmprasath commented on spring-cloud-aws-cloudmap-config/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapRegistryService.java:
```diff
@@ -0,0 +1,324 @@
+
+	@Override
+	public void onApplicationEvent(ContextClosedEvent event) {
+		this.deregisterInstance();
```
Updated

hariohmprasath commented on spring-cloud-aws-cloudmap-config/src/test/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapAnnotationScannerTest.java:
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
```
Updated

hariohmprasath commented:
> @hariohmprasath thank you so much for the PR! I am really excited about this one TBH and I don't know why. I have submitted some comments to help you organize the modules much better. Please, don't forget to add yourself as an author for the classes that have been touched.
> @eddumelendez Thanks a lot for all the support and helping with the review. Really appreciate it. I have responded to your comments and I am happy to answer any further questions you may have.


> I'm not an expert in cloudmap and left one question to be clear about it. Also, it would be nice to add some javadocs to know more the class responsibility.
> Sure I have added more javadocs. Once the changes are approved, I'm planning to create a blog post with end to end deployment scripts using AWS CDK (which can build a ECS cluster and deploying this application) so it would be easier for anyone to incorporate this in future.

MatejNedic commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapDiscovery.java:
```diff
@@ -0,0 +1,54 @@
+ */
+public class CloudMapDiscovery {
+
+	private boolean failFast;
```
I think we should default `failFast`  to true.

hariohmprasath replied:
I agree. I have updated the code to make it fail (by default) if discovery has failed

hariohmprasath commented:
Hi @spencergibb, @eddumelendez, @MatejNedic, @jkuipers,

Hope everyone had a wonderful weekend. I have incorporated everyone's feedback, just want to check whether we are fine in merging this pull request or do we have any feedback that you would like to incorporate. Happy to do so.

Let me know, appreciate all the love and support from this community. Thanks.

vanekjar commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapDiscoverService.java:
```diff
@@ -0,0 +1,68 @@
+
+		if (properties.getFilterAttributes() != null && !properties.getFilterAttributes().isEmpty()) {
+			Map<String, String> filterMap = new HashMap<>();
+			for (String key : properties.getFilterAttributes().keySet()) {
+				filterMap.put(key, properties.getFilterAttributes().get(key));
+			}
```
I don't think this is the best way how to iterate `Map` in Java. The following was may be better:

```java
for (Map.Entry<String, String> entry : map.entrySet()) {
    System.out.println(entry.getKey() + "/" + entry.getValue());
}
```

hariohmprasath replied:
Thanks. Its updated now

vanekjar commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapDiscoveryProperties.java:
```diff
@@ -0,0 +1,71 @@
+ */
+public class CloudMapDiscoveryProperties {
+
+	private String serviceNameSpace;
```
`NameSpace` is usually just a single word

hariohmprasath replied:
I have updated all the `serviceNameSpace` names to `nameSpace` to keep it consistent

vanekjar commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapRegistryService.java:
```diff
@@ -0,0 +1,370 @@
+					.withNamespaceId(nameSpaceId).withDnsConfig(
+							new DnsConfig().withDnsRecords(new DnsRecord().withType(RecordType.A).withTTL(300L)));
+
+			if (!StringUtils.isNullOrEmpty(properties.getHealthCheckResourcePath())) {
+				HealthCheckConfig healthCheckConfig = new HealthCheckConfig();
+				healthCheckConfig.setType(String.valueOf(("https").equalsIgnoreCase(properties.getHealthCheckProtocol())
+						? HealthCheckType.HTTPS : HealthCheckType.HTTP));
+				healthCheckConfig.setResourcePath(properties.getHealthCheckResourcePath());
+				healthCheckConfig.setFailureThreshold(
+						(properties.getHealthCheckThreshold() == null) ? 5 : properties.getHealthCheckThreshold());
+				serviceRequest.withHealthCheckConfig(healthCheckConfig);
+			}
```
Health checks are supported only for `public` or `HTTP` namespaces. It doesn't work for `private` namespace that are used here.

https://docs.aws.amazon.com/cloud-map/latest/api/API_CreateService.html

hariohmprasath replied:
Good catch. I have removed these additional methods and code

vanekjar reviewed (changes requested):
Thanks for the integration with AWS Cloud Map. I added a few comments.

hariohmprasath commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapPropertySourceLocator.java:
```diff
@@ -0,0 +1,78 @@
+
+	private final CloudMapDiscoverService instanceDiscovery;
+
+	public AwsCloudMapPropertySourceLocator(AWSServiceDiscovery serviceDiscovery, CloudMapDiscovery cloudMapDiscovery,
+			CloudMapDiscoverService instanceDiscovery) {
```
Added javadoc on top of the method, explaining what each of these object refers to

```java
/**
* Constructor method.
* @param serviceDiscovery AWS CloudMap service discovery
* @param cloudMapDiscovery CloudMap discovery POJO (user specified)
* @param instanceDiscovery Service class for cloud map registration
*/
```

hariohmprasath commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapPropertySources.java:
```diff
@@ -0,0 +1,61 @@
+
+import com.amazonaws.services.servicediscovery.AWSServiceDiscovery;
+
+/**
+ * Recursively retrieves all Http instances based on cloudmap namespace and services from
+ * the AWS CloudMap using the provided AWS ServiceDiscovery client.
```
Updated the comment to describe what the method does

hariohmprasath commented:
> Thanks for the integration with AWS Cloud Map. I added a few comments.

Thanks @vanekjar for all the feedback. I have incorporated all the changes.

vanekjar commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapRegistryService.java:
```diff
@@ -0,0 +1,355 @@
+	 * with the created namespace and service
+	 * @return cloudmap registration operation ID
+	 */
+	public String registerInstances() {
+		if (properties != null && !StringUtils.isNullOrEmpty(properties.getNameSpace())
+				&& !StringUtils.isNullOrEmpty(properties.getService())) {
+
+			final String nameSpace = properties.getNameSpace();
+			final String service = properties.getService();
+
+			this.serviceInstanceId = UUID.randomUUID().toString();
```
Is it the case that there is always a single instance per service? I am surprised that `registerInstances` actually registers a single instance only.

hariohmprasath replied:
Let's say we have two tasks running per service, this piece of code will create one Cloudmap service and registers two HTTP instances inside the service. There can be only one service and multiple HTTP instances gets registered inside that.

vanekjar replied:
I see. I wondered why there is `registerInstances` (plural) and `deregisterInstance` `singular`. That sounds like instances may be left behind. But now I see it's actually just a naming issue.

hariohmprasath replied:
To avoid further confusion, I have renamed the method to `registerInstance()`

hariohmprasath commented:
Hi @eddumelendez & @vanekjar,
I have taken care of all the requested changes, so following up on this PR to see whether this can get merged to the repository?

Thanks
Hari

eddumelendez commented:
@hariohmprasath sorry for the delay. I will review on Thursday but still see things to improve. I think we should see at how https://github.com/spring-cloud/spring-cloud-consul/ has been implemented. More details on Thursday.

hariohmprasath commented:
> @hariohmprasath sorry for the delay. I will review on Thursday but still see things to improve. I think we should see at how https://github.com/spring-cloud/spring-cloud-consul/ has been implemented. More details on Thursday.

Thanks, appreciate the help on this

MatejNedic commented on spring-cloud-aws-cloudmap/src/test/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapRegisterServiceTest.java:
```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
```
2021

hariohmprasath replied:
Done

MatejNedic commented on spring-cloud-aws-cloudmap/src/test/java/org/springframework/cloud/aws/cloudmap/CloudMapTestUtils.java:
```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
```
2021

hariohmprasath replied:
Done

MatejNedic reviewed (changes requested):
Hey @hariohmprasath , I left few comments about changes.

eddumelendez commented:
@hariohmprasath please make sure this PR only contains your changes. it is hard to review.

hariohmprasath commented:
> @hariohmprasath please make sure this PR only contains your changes. it is hard to review.

@eddumelendez this PR only contains changes related to cloudmap integration. We got lot of good feedback from this community and everything related to this feature has been incorporated part of this PR.

Thanks
Hari

eddumelendez commented:
check the commits tab, please

hariohmprasath commented on spring-cloud-starter-aws-cloudmap/src/main/java/org/springframework/cloud/aws/autoconfigure/cloudmap/AwsCloudMapBootstrapConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+@EnableConfigurationProperties(CloudMapProperties.class)
+@ConditionalOnClass({ AWSServiceDiscovery.class })
+@ConditionalOnProperty(prefix = CloudMapProperties.CONFIG_PREFIX, name = "enabled", matchIfMissing = true)
+@ComponentScan
```
Done

hariohmprasath commented on spring-cloud-starter-aws-cloudmap/src/main/java/org/springframework/cloud/aws/autoconfigure/cloudmap/AwsCloudMapBootstrapConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+ * Cloudmap BootstrapConfiguration configuration class to create the required beans.
+ *
+ * @author Hari Ohm Prasath
+ * @since 1.0
```
Done

hariohmprasath commented:
> check the commits tab, please

There were other changes that were happening in this branch in parallel while this pull request was open. So I have to rebase my changes with the target branch to make it compatible for getting this merged, which has led to other commits getting in the way of this one. If you remember I infact created a discussion to ask a few questions around this. 

What do you suggest? I am open to make things easier for you to review so I can take care of the comments. Let me know your thoughts on this. Thanks.

eddumelendez commented:
after the rebase is done only your commits should be displayed. looks like in the process some commits from other authors keep in your branch. there is no need to update the PR with changes happening in the main branch but this time I will appreciate to do it and only keep your commits, please

hariohmprasath commented:
> after the rebase is done only your commits should be displayed. looks like in the process some commits from other authors keep in your branch. there is no need to update the PR with changes happening in the main branch but this time I will appreciate to do it and only keep your commits, please

Sorry i M 

> after the rebase is done only your commits should be displayed. looks like in the process some commits from other authors keep in your branch. there is no need to update the PR with changes happening in the main branch but this time I will appreciate to do it and only keep your commits, please

Are you fine in reviewing it this time? or do you want me to change the PR. Sorry I am trying to understand your response. Appreciate your quick responses. Thanks

eddumelendez commented:
@hariohmprasath please, just make sure the branch contains only your commits related to cloudmap integration

hariohmprasath commented:
> please, just make sure the branch contains only your commits related to cloudmap integration

Hi @eddumelendez,
I might need some help on this. Only way I can think of doing this is creating a new fork/branch and merging the cloudmap changes that we have in this pull request on to the new branch. Do you want me to that?

Or do we have a way of doing what you have in this current pull request? I know we are almost there with a lot of reviews done and feedback incorporated frankly I am not an expert with GIT so can you please help with this? Any pointer to how I can do this on the current one will be really helpful.

Thanks in advance

- Hari

hariohmprasath commented:
Hi @eddumelendez & @spencergibb  ,
Thanks to @juho9000 I finally cleaned up this pull request with just my commits. Will you be able to review it now? If you feedback drop a comment, I can address them.

Thanks
Hari

hariohmprasath commented:
Hi @eddumelendez
Will you be able to review this pull request? Currently it just contains my changes.

juho9000 commented:
Is there still something that needs to be done to get this merged? @eddumelendez

juho9000 commented:
It's been a long time since the required changes were submitted, any chance of reviewing? Eagerly waiting for these changes to be merged! Pinging @eddumelendez @maciejwalkowiak

hariohmprasath commented:
Hi @eddumelendez and @maciejwalkowiak,
This PR can serve as a substitute for customers using `eurkera` and moving into AWS. If you have any suggestions or recommendations on the PR please let me know, happy to work on it. Thanks.

cyky commented:
Is there more work needed to get this merged @maciejwalkowiak and @eddumelendez? It would be really helpful to get this feature merged soon.

willome commented:
Any news for this PR ?

migroskub commented:
Any news?

hariohmprasath commented:
Hi @eddumelendez & @maciejwalkowiak,
Can you please respond to this PR? I see a lot of interest around this. If you still feel this is not the right place to get it merged, that's fine too, but please respond so we can either close this PR or merge it to the main branch.

 Hari

eddumelendez commented:
@hariohmprasath first of all, apologies for the long absent. We are still interested in this feature and I left a comment to something similar to [spring-cloud-consul-discovery-module](https://github.com/spring-cloud/spring-cloud-consul/tree/main/spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul) which will fit with what CloudMap provides.

kdhrubo commented:
We also need this feature urgently. When can this merged to a release. Is this compatible with with Spring Boot 2.6.x?

eddumelendez commented:
@kdhrubo currently, @hariohmprasath is working on some changes to make it work. This a contribution for an external member of the team coding of his spare time and we don't have a target date for this feature.

kdhrubo commented:
@eddumelendez - i understand. how can i help to get this merged? 

Also will this also support service to service load balancing with spring cloud load balancer and feign?

eddumelendez commented:
Indeed, our suggestion is [this one](https://github.com/awspring/spring-cloud-aws/pull/95#discussion_r804973168) which as far as I remember use the load balancer. Not sure about feign tho, IMHO there is no development needed on this side for that

kdhrubo commented:
Sounds good @eddumelendez . Do you need testing help? Given that dev is completed - any reviews pending?

hariohmprasath commented:
Hi @kdhrubo,
Based on the suggestion from @eddumelendez. I rewrote this module using spring discovery. You can find an early draft of this PR here:

[https://github.com/hariohmprasath/spring-cloud-aws/commit/4339b0a6d32bf8c7e5248f93fb6f0d3dc8edae90

I have reached out @eddumelendez and team for some early peer review before i wrap up the testcases and submit a PR

kdhrubo commented:
@hariohmprasath thanks. I will try to test the code this week and share my observation

hariohmprasath commented:
> @hariohmprasath thanks. I will try to test the code this week and share my observation. Awesome 👍 work.

Thanks @kdhrubo. You can directly reach out to me at this email address hariohmprasath@gmail.com

MatejNedic reviewed (changes requested):
Hey @hariohmprasath, tnx for your contribution!
It is looking good so far.

eddumelendez commented on spring-cloud-aws-cloudmap/pom.xml:
```diff
@@ -0,0 +1,60 @@
+		<dependency>
+			<groupId>net.oneandone.reflections8</groupId>
+			<artifactId>reflections8</artifactId>
+			<version>0.11.7</version>
```
I wonder where it is used or if it is needed for another dependency

hariohmprasath replied:
I have removed this as it is no longer needed

eddumelendez commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/http/CloudMapRequestFactory.java:
```diff
@@ -0,0 +1,18 @@
+import java.net.URI;
+
+@Service
+public class CloudMapRequestFactory implements ClientHttpRequestFactory {
```
it should be define in the auto-configure module, right? not in the starter

hariohmprasath replied:
I have removed this as its not getting used

eddumelendez commented on spring-cloud-starter-aws-cloudmap/src/main/java/org/springframework/cloud/aws/autoconfigure/cloudmap/AwsCloudMapBootstrapConfiguration.java:
```diff
@@ -0,0 +1,90 @@
+ */
+@Configuration(proxyBeanMethods = false)
+@EnableConfigurationProperties(CloudMapProperties.class)
+@ConditionalOnClass({AWSServiceDiscovery.class})
```
those 3 lives in the same library, one of them is just needed in the `@ConditionalOnClass`. We can add `org.springframework.cloud.client.discovery.event.InstanceRegisteredEvent` and `org.springframework.cloud.client.serviceregistry.AutoServiceRegistration` if live in different libraries.

hariohmprasath replied:
I have added the missing beans as part of the `@ConditionalOnClass` annotations.  But not sure about `InstanceRegisteredEvent` and `AutoServiceRegistration` as I am not using it.

eddumelendez commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapUtils.java:
```diff
@@ -0,0 +1,457 @@
+			ResponseEntity<String> metaDataResponse = REST_TEMPLATE
+				.getForEntity(System.getenv("ECS_CONTAINER_METADATA_URI_V4") + "/task", String.class);
+			JsonNode root = JSON_MAPPER.readTree(metaDataResponse.getBody());
+			JsonNode jsonNode = root.get("Containers").get(0).get("Networks").get(0);
```
is this always true? I mean the first item in the container list and the first network? In general I think is ok but just double checking

hariohmprasath replied:
Good catch. I have created two methods which will be picked up based on the `DEPLOYMENT_PLATFORM` value `ECS`

* `getEcsRegistrationAttributes` - Used for ECS
* `getEksRegistrationAttributes` - Used for EKS

eddumelendez commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapUtils.java:
```diff
@@ -0,0 +1,457 @@
+		Map<String, String> attributes = new HashMap<>();
+		try {
+			ResponseEntity<String> metaDataResponse = REST_TEMPLATE
+				.getForEntity(System.getenv("ECS_CONTAINER_METADATA_URI_V4") + "/task", String.class);
```
the system environment variable is for ECS only right? ehat about services in EC2 or EKS?

hariohmprasath replied:
Now we have two different methods to handle ECS, EKS. Inside ECS, we have a mechanism to use default metaURL if not available (to handle EC2 use case)

hariohmprasath commented on spring-cloud-starter-aws-cloudmap/src/main/java/org/springframework/cloud/aws/autoconfigure/cloudmap/AwsCloudMapBootstrapConfiguration.java:
```diff
@@ -0,0 +1,90 @@
+	@Bean
+	@ConditionalOnMissingBean
+	AWSServiceDiscovery serviceDiscovery(CloudMapProperties properties) {
+		return createServiceDiscoveryClient(properties);
```
Yes, you are right. Removed it in the new version

hariohmprasath commented on spring-cloud-starter-aws-cloudmap/src/main/java/org/springframework/cloud/aws/autoconfigure/cloudmap/AwsCloudMapBootstrapConfiguration.java:
```diff
@@ -0,0 +1,90 @@
+public class AwsCloudMapBootstrapConfiguration {
+
+	@Autowired
+	private ApplicationContext context;
```
Resolved

hariohmprasath commented on spring-cloud-starter-aws-cloudmap/src/main/java/org/springframework/cloud/aws/autoconfigure/cloudmap/AwsCloudMapBootstrapConfiguration.java:
```diff
@@ -0,0 +1,90 @@
+@ConditionalOnDiscoveryEnabled
+@ConditionalOnBlockingDiscoveryEnabled
+public class AwsCloudMapBootstrapConfiguration {
+
```
Fixed

hariohmprasath commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapUtils.java:
```diff
@@ -0,0 +1,457 @@
+	public final String SERVICE_ID = "SERVICE_ID";
+	public final String SERVICE_INSTANCE_ID = "SERVICE_INSTANCE_ID";
+	public final String IPV_4_ADDRESS = "IPV4_ADDRESS";
+	final RestTemplate REST_TEMPLATE = new RestTemplate();
```
Makes sense, I have updated the code

hariohmprasath commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapUtils.java:
```diff
@@ -0,0 +1,457 @@
+ * @author Hari Ohm Prasath
+ * @since 2.3.2
+ */
+public enum CloudMapUtils {
```
Yes keeping it as Singleton using enum, compared better compared to static methods

hariohmprasath commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapUtils.java:
```diff
@@ -0,0 +1,457 @@
+	public Map<String, String> getRegistrationAttributes() {
+		Map<String, String> attributes = new HashMap<>();
+		try {
+			ResponseEntity<String> metaDataResponse = REST_TEMPLATE
```
Updated it based on WEB_CLIENT

hariohmprasath commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapUtils.java:
```diff
@@ -0,0 +1,457 @@
+						final ServiceFilter serviceFilter = new ServiceFilter().withName(NAMESPACE_ID)
+							.withCondition("EQ").withValues(nameSpaceId);
+						final ListServicesRequest servicesRequest = new ListServicesRequest().withFilters(serviceFilter);
+						if (token != null)
```
Code updated

hariohmprasath commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapUtils.java:
```diff
@@ -0,0 +1,457 @@
+
+			List<NamespaceSummary> namespaceSummaries = namespacesResult.getNamespaces();
+			Optional<String> namespaceId = namespaceSummaries.stream()
+				.filter(n -> n.getName().equals(nameSpace))
```
Yes thats correct

hariohmprasath commented:
Hi @eddumelendez, @MatejNedic,

Thanks for the quick response on the code review and for helping with the early feedback. It was super helpful. I have updated the code based on the feedback, additionally tested the integration both in `ECS` and `EKS`, with default behavior for `EC2`. It's working fine. Along with this, I also added unit test cases to validate the code. Please review it and let me know if you have questions, appreciate all the help and support for the team.

eddumelendez commented on spring-cloud-starter-aws-cloudmap/src/main/java/org/springframework/cloud/aws/autoconfigure/cloudmap/AwsCloudMapBootstrapConfiguration.java:
```diff
@@ -0,0 +1,89 @@
+@ConditionalOnProperty(prefix = CloudMapProperties.CONFIG_PREFIX, name = "enabled", matchIfMissing = true)
+@ConditionalOnDiscoveryEnabled
+@ConditionalOnBlockingDiscoveryEnabled
+public class AwsCloudMapBootstrapConfiguration {
```
```suggestion
public class CloudMapBootstrapConfiguration {
```

hariohmprasath replied:
Updated

eddumelendez commented on spring-cloud-starter-aws-cloudmap/src/main/java/org/springframework/cloud/aws/autoconfigure/cloudmap/AwsCloudMapBootstrapConfiguration.java:
```diff
@@ -0,0 +1,89 @@
+
+	public AwsCloudMapBootstrapConfiguration(CloudMapProperties properties, ApplicationContext context) {
+		AWSServiceDiscoveryClientBuilder builder = AWSServiceDiscoveryClientBuilder.standard()
+				.withCredentials(new DefaultAWSCredentialsProviderChain());
```
can you add `.overrideConfiguration(SpringCloudClientConfiguration.clientOverrideConfiguration())` and `Optional.ofNullable(this.properties.getEndpoint()).ifPresent(client::endpointOverride);` similar to other auto-configurations. Also. Region and Credentials can be injected. See other auto-configurations

hariohmprasath replied:
Fixed the overrides

eddumelendez reviewed (commented):
thanks for the update @hariohmprasath ! I have added some suggestions and questions in order to understand a little bit more about the implementation. 

Please, take into account the following points
* starter should only contain a pom
* cloudmap module should not have dependency on spring-boot. So far I have identified it depends on `ConfigurationProperties`
* auto-configuration is only managed at `spring-cloud-aws-autoconfigure `module

Vostan commented:
Hi, we really seeking this module to reach the main branch, is there something we can do to boost this up?

nstdio commented:
Thanks for the great effort! We are looking forward to see this PR merged. Is there anything that we can possibly help?

hariohmprasath commented on spring-cloud-aws-autoconfigure/src/main/resources/META-INF/spring.factories:
```diff
@@ -12,3 +12,5 @@ io.awspring.cloud.autoconfigure.messaging.SnsAutoConfiguration,\
 io.awspring.cloud.autoconfigure.metrics.CloudWatchExportAutoConfiguration,\
 io.awspring.cloud.autoconfigure.security.CognitoAuthenticationAutoConfiguration
+org.springframework.cloud.client.discovery.EnableDiscoveryClient=\
+org.springframework.cloud.aws.cloudmap.servicediscovery.CloudMapDiscoveryClient
```
You are right, this is not required. So I have removed it.

hariohmprasath replied:
Hi @eddumelendez,
Please find my responses below:
* starter should only contain a pom & auto-configuration is only managed at spring-cloud-aws-autoconfigure module
> I followed `spring-cloud-starter-aws-parameter-store-config` module implementation to begin with and since we have write a discovery and auto-registration client part of `spring-cloud-aws-cloudmap` module I am unable to move the autoconfiguration to `autoconfigure` module.

* cloudmap module should not have dependency on spring-boot. So far I have identified it depends on ConfigurationProperties
> Removed `spring-boot` dependencies from the module

hariohmprasath commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapUtils.java:
```diff
@@ -0,0 +1,600 @@
+	 */
+	INSTANCE;
+
+	public static final String META_DATA_URL = "http://169.254.169.254/latest/meta-data";
```
Updated variable name

hariohmprasath commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapUtils.java:
```diff
@@ -0,0 +1,600 @@
+	public Map<String, String> getRegistrationAttributes() {
+		String deploymentPlatform = System.getenv(DEPLOYMENT_PLATFORM);
+		LOGGER.info("Deployment platform passed in {}", deploymentPlatform);
+		if (StringUtils.hasText(deploymentPlatform) && EKS.equalsIgnoreCase(deploymentPlatform.trim()))
```
Updated `CloudMapUtils` to a singleton class and fixed space issue

hariohmprasath commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapUtils.java:
```diff
@@ -0,0 +1,600 @@
+	 * register instances to cloudmap service.
+	 * @return map containing ip address and vpcid
+	 */
+	public Map<String, String> getRegistrationAttributes() {
```
I have added additional docs to clarify this, let me know if its still not clear

hariohmprasath commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/model/discovery/CloudMapDiscovery.java:
```diff
@@ -0,0 +1,55 @@
+	private List<CloudMapDiscoveryProperties> discoveryList;
+
+	public boolean isFailFast() {
+		return failFast;
```
Fixed

hariohmprasath commented on spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/model/discovery/CloudMapDiscovery.java:
```diff
@@ -0,0 +1,55 @@
+	}
+
+	public List<CloudMapDiscoveryProperties> getDiscoveryList() {
+		return discoveryList;
```
Fixed

hariohmprasath commented on spring-cloud-aws-cloudmap/src/test/java/org/springframework/cloud/aws/cloudmap/AwsCloudMapRegisterServiceTest.java:
```diff
@@ -0,0 +1,205 @@
+ * @author Hari Ohm Prasath
+ * @since 2.3.2
+ */
+public class AwsCloudMapRegisterServiceTest {
```
Updated

hariohmprasath commented:
Hi @eddumelendez,
Sorry about my delay in submitting the updates to the PR. Finally I got a chance to take care of all the requested changes, please review it and let me know if you have any questions. Thanks

Vostan commented:
@maciejwalkowiak If you would have a chance to review, these changes and merge them that would be amazing.

maciejwalkowiak commented:
I am a bit puzzled right now because since this PR started we got huge progress with migration to AWS SDK v2 in Spring Cloud AWS 3.0.

The options we have are: merge CloudMap to 2.x and release 3.0 without CloudMap with an idea to migrate it to AWS SDK v2 in 3.1, or rebase this PR on the top of `main` branch migrated to AWS SDK v2 and include it in 3.x at one point - but this would postpone having CloudMap support even further.

@hariohmprasath are you able to estimate the efforts needed to migrate this PR to AWS SDK v2 and if you would be even willing to do it?

Vostan commented:
> I am a bit puzzled right now because since this PR started we got huge progress with migration to AWS SDK v2 in Spring Cloud AWS 3.0.
> 
> The options we have are: merge CloudMap to 2.x and release 3.0 without CloudMap with an idea to migrate it to AWS SDK v2 in 3.1, or rebase this PR on the top of `main` branch migrated to AWS SDK v2 and include it in 3.x at one point - but this would postpone having CloudMap support even further.
> 
> @hariohmprasath are you able to estimate the efforts needed to migrate this PR to AWS SDK v2 and if you would be even willing to do it?

Is there some way I can help @hariohmprasath ?

hariohmprasath commented:
> I am a bit puzzled right now because since this PR started we got huge progress with migration to AWS SDK v2 in Spring Cloud AWS 3.0.
> 
> The options we have are: merge CloudMap to 2.x and release 3.0 without CloudMap with an idea to migrate it to AWS SDK v2 in 3.1, or rebase this PR on the top of `main` branch migrated to AWS SDK v2 and include it in 3.x at one point - but this would postpone having CloudMap support even further.
> 
> @hariohmprasath are you able to estimate the efforts needed to migrate this PR to AWS SDK v2 and if you would be even willing to do it?

Hi @maciejwalkowiak,
Thanks for going over the options and I can completely understand the timelines involved in releasing 3.x with AWS SDK v2 support. So I took some time over this weekend to rework this feature based on AWS SDK 2.x. I have submitted a [new PR (for 3.x branch)](https://github.com/awspring/spring-cloud-aws/pull/506) with all the required changes, tested it using AWS ECS and everything seems to work fine. Since this PR got already reviewed by the team, the new PR based on AWS SDK 2.x contains all the requested changes. Hopefully, this will make things easier on your end for planning the release and let me know if you have questions or want me to take care of anything before merging this change. Thanks again for the support from this community.

https://github.com/awspring/spring-cloud-aws/pull/506

hariohmprasath commented:
> > I am a bit puzzled right now because since this PR started we got huge progress with migration to AWS SDK v2 in Spring Cloud AWS 3.0.
> > The options we have are: merge CloudMap to 2.x and release 3.0 without CloudMap with an idea to migrate it to AWS SDK v2 in 3.1, or rebase this PR on the top of `main` branch migrated to AWS SDK v2 and include it in 3.x at one point - but this would postpone having CloudMap support even further.
> > @hariohmprasath are you able to estimate the efforts needed to migrate this PR to AWS SDK v2 and if you would be even willing to do it?
> 
> Is there some way I can help @hariohmprasath ?

Thanks @Vostan for volunteering to help me out on this. Since we have a new PR posted based on AWS SDK 2.x we should be good in releasing this feature on time. I would definitely need your help in trying this feature out and work on improving the integration with AWS CloudMap. Thanks

hariohmprasath commented:
Hi Maciej,

Thanks for going over the options and I can completely understand the
timelines involved in releasing 3.x with AWS SDK v2 support. So I took some
time over this weekend to rework this feature based on AWS SDK 2.x. I have
submitted a new PR (for 3.x branch)
<https://github.com/awspring/spring-cloud-aws/pull/506> with all the
required changes, tested it using AWS ECS and everything seems to work
fine. Since this PR has already been reviewed by the team, the new PR based
on AWS SDK 2.x contains all the requested changes. Hopefully, this will
make things easier on your end for planning the release and let me know if
you have questions or want me to take care of anything before merging this
change. Thanks again for all the support from this team.

#506 <https://github.com/awspring/spring-cloud-aws/pull/506>

On Thu, Sep 1, 2022 at 6:37 AM Maciej Walkowiak ***@***.***>
wrote:

> I am a bit puzzled right now because since this PR started we got huge
> progress with migration to AWS SDK v2 in Spring Cloud AWS 3.0.
>
> The options we have are: merge CloudMap to 2.x and release 3.0 without
> CloudMap with an idea to migrate it to AWS SDK v2 in 3.1, or rebase this PR
> on the top of main branch migrated to AWS SDK v2 and include it in 3.x at
> one point - but this would postpone having CloudMap support even further.
>
> @hariohmprasath <https://github.com/hariohmprasath> are you able to
> estimate the efforts needed to migrate this PR to AWS SDK v2 and if you
> would be even willing to do it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/awspring/spring-cloud-aws/pull/95#issuecomment-1234293750>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAT6QHJJSULQHQ2N2HSPO6LV4CWRHANCNFSM4YSKINKA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
With Regards,
Hari Ohm Prasath

maciejwalkowiak commented:
@hariohmprasath fantastic! Thanks a lot for putting effort into this! I'm scheduling CloudMap support for M3 and do our best to review and try it out on time.

maciejwalkowiak commented:
Closing in favour of #506
